pandas/tseries/index.py

```diff
@@ -1637,6 +1637,24 @@ def delete(self, loc):
             new_dates = tslib.tz_convert(new_dates, 'UTC', self.tz)
         return DatetimeIndex(new_dates, name=self.name, freq=freq, tz=self.tz)
 
+    def putmask(self, mask, value):
+        """
+        return a new Index of the values set with the mask
+
+        See also
+        --------
+        numpy.ndarray.putmask
+        """
+        zzone = tslib.get_timezone(self.tzinfo)
+        vzone = tslib.get_timezone(getattr(value, 'tzinfo', None))
+        if zzone == vzone:
+            try:
+                value = tslib.Timestamp(value).asm8
+                return super(DatetimeIndex, self).putmask(mask, value)
+            except (TypeError, ValueError):
```

Comment:
Currently no to handle value without ``tzinfo`` & ``DatetimeIndex`` without tz.

Changing to return dummy ``tslib.get_timezone(getattr(value, 'tzinfo', '__no_tz__'))`` not to match with ``DatetimeIndex`` without tz.